hal/src/common/thumbv7em/clock/v2/gclk.rs

```diff
@@ -359,6 +364,56 @@ where
     }
 }
 
+impl<T: GclkSourceType> Gclk<Gen0, T, One> {
+    /// TODO
+    #[inline]
+    pub unsafe fn swap<Old, New>(
+        self,
+        old: Old,
+        new: New,
+    ) -> (Gclk<Gen0, New::Type, One>, Old::Unlocked, New::Locked)
+    where
+        Old: GclkSource<Gen0, Type = T> + Unlockable,
+        New: GclkSource<Gen0> + Lockable,
+    {
+        let (config, old, new) = self.config.swap(old, new);
+        (Gclk::create(config, self.count), old, new)
+    }
+}
+
+impl Gclk<Gen0, marker::Dfll<OpenMode>, One> {
+    pub unsafe fn change_mode<T: PclkSourceType>(
+        self,
+        old: Dfll<OpenMode, One>,
+        exchange: impl FnOnce(Dfll<OpenMode>) -> Dfll<ClosedMode<T>>,
+    ) -> (
+        Gclk<Gen0, marker::Dfll<marker::ClosedMode>, One>,
+        Dfll<ClosedMode<T>, One>,
+    ) {
+        let (token, old) = self.config.free(old);
+        let new = exchange(old);
+        let (config, new) = GclkConfig::new(token, new);
+        (Gclk::create(config, self.count), new)
+    }
+}
+
+impl Gclk<Gen0, marker::Dfll<marker::ClosedMode>, One> {
```

Comment:
This might be a little bit of an overkill but I let it be for now.

Reply:
@bradleyharden I could avoid keeping mode in a source/marker type if I could get away with a more general specialization for `Dfll` and have a generic closure. I went this route because I wanted a way to retrieve `Pclk0` in case user changes mode from `ClosedLoop` to `OpenLoop`. The other way around it is perfectly fine, as one can just move `Pclk0` to closure. Moving out of closure is not possible.. I think.